File: 1913. Maximum Product Difference Between Two Pairs/Solution.py
from typing import List


class Solution:
    def maxProductDifference(self, nums: List[int]) -> int:
        mx1 = mn2 = max(nums[0], nums[1])
        mx2 = mn1 = min(nums[0], nums[1])
        for x in nums[2:]:
            if x >= mx1:
                mx2 = mx1
                mx1 = x
            elif x > mx2:
                mx2 = x
            if x <= mn1:
                mn2 = mn1
                mn1 = x
            elif x < mn2:
                mn2 = x
        return mx1 * mx2 - mn1 * mn2

File: 1913. Maximum Product Difference Between Two Pairs/Solution.java
class Solution {
    public int maxProductDifference(int[] nums) {
        int max1 = Math.max(nums[0], nums[1]);
        int max2 = Math.min(nums[0], nums[1]);
        int min1 = max2, min2 = max1;
        for (int i = 2; i < nums.length; i++) {
            if (nums[i] >= max1) {
                max2 = max1;
                max1 = nums[i];
            } else if (nums[i] > max2) {
                max2 = nums[i];
            }
            if (nums[i] <= min1) {
                min2 = min1;
                min1 = nums[i];
            } else if (nums[i] < min2) {
                min2 = nums[i];
            }
        }
        return max1 * max2 - min1 * min2;
    }
}
